lib/xml/parser.js

```diff
@@ -178,48 +176,51 @@ function parseEvents(xml) {
                     dtstart: vevent.getFirstPropertyValue("dtstart"),
                 });
+                const recur = vevent.getFirstPropertyValue("rrule");
+                // Expand upto a maximum of 10 upcoming occurrences for events that never end
+                const MAX_INFINITE_OCCURRENCES_COUNT = 10;
+                const startRange = startDate ? ICAL.Time.fromDateTimeString(moment(startDate).toISOString()) : null;
+                const endRange = endDate ? ICAL.Time.fromDateTimeString(moment(endDate).toISOString()) : null;
+                let occurrencesCount = 0;
+
+                for (let next = expand.next(); next; next = expand.next()) {
+                    if (startRange && next.compare(startRange) < 0) {
+                        continue;
+                    }
+
+                    if (endRange && next.compare(endRange) > 0) {
+                        break;
+                    }
 
-                // Since there are infinite rules, its a good idea to limit the scope
-                // of the iteration then resume later on
-                // Expand upto a maximum of 10 upcoming occurances
-                for (let i = 0; i < 10; i++) {
-                    const nextOccuranceTime = expand.next();
-
-                    if (!expand.complete) {
-                        // Handle this next expanded occurence
-                        const nextEvent = icalEvent.getOccurrenceDetails(nextOccuranceTime);
-
-                        eventData = {};
-
-                        if (modifiedOccurences.length === 0) {
-                            // No events have been modified
-                            formatted.push({
-                                ics: event.href[0],
-                                etag,
-                                data: getNormalOccurenceEventData(nextEvent, eventData, vevent),
-                            });
-                        } else if (isModified(nextOccuranceTime, modifiedOccurences)) {
-                            // This is the event that has been modied
-                            const key = getModifiedOccuranceKey(nextOccuranceTime, modifiedOccurences) || 0;
-
-                            formatted.push({
-                                ics: event.href[0],
-                                etag,
-                                data: getModifiedOccurenceEventData(vevents[key], eventData),
-                            });
-                        } else {
-                            // Expand this event normally
-                            formatted.push({
-                                ics: event.href[0],
-                                etag,
-                                data: getNormalOccurenceEventData(nextEvent, eventData, vevent),
-                            });
-                        }
+                    if (!endRange && !recur.isFinite() && occurrencesCount >= MAX_INFINITE_OCCURRENCES_COUNT) {
+                        break;
+                    }
```

Comment:
Thanks for raising the almost infinite loop possibility :thumbsup: The only case where I see an almost infinite loop occurring is when an event is finite but has an end date (or count) that is way too far in the future (for infinite events we are already capping the iterations to 10). Are we seeing this the same way?

I like the idea of the library user specifying the upper iteration bound :thumbsup: If I understand you well, I will include `maxExpandCount` in the config options provided by the user as a required parameter and pass it to all events:

```js
const config = {
    auth: {
        user: BAIKAL_USER,
        pass: BAIKAL_PASSWORD,
    },
    uri: BAIKAL_URI,
    maxExpandCount: 10,
};
```

> Furthermore I think there's an opportunity for improvement here: we could remove all recurring events that start after the specified end date. That would save us some iterations.

This situation does not exist. When we query CalDav with an end date specified, the response we get only includes recurring events that start (or have an occurrence) between the startRange and endRange.

> If they don't care about this issue, they still can pass maxExpandCount: Infinity.

Here is where It gets blurry again: (going with the logic you suggested)

CASE 1: startRange and endRange provided and maxExpandCount: Infinity

-   If the endRange is near the startRange, then all is well. However, if the endRange is way into the future, we have the same possibility of an almost infinite loop, only that the user opted into it - this is OK

CASE 2: startRange provided, endRange not provided and maxExpandCount: Infinity

-   What happens here? This will loop indefinitely... (at least until the year 10000 breaks the JS Date :smile:) I do not see a use case of maxExpandCount being set to Infinity, **it should be a finite number always**.
What do you think?

> Maybe you can also add an error if startRange is after endRange.

Right now we silently change the endRange to `null` if this case happens but I agree with throwing an error :thumbsup:

Reply:
> Thanks for raising the almost infinite loop possibility 👍 The only case where I see an almost infinite loop occurring is when an event is finite but has an end date (or count) that is way too far in the future (for infinite events we are already capping the iterations to 10). Are we seeing this the same way?

Yes, I was referring to this case. I don't know how common that is. But I know that some tools limit the max iteration count or the end date when creating new re-occurring events.

I think I would scope the config option under `events`:

```ts
const config = {
    auth: {
        user: BAIKAL_USER,
        pass: BAIKAL_PASSWORD,
    },
    uri: BAIKAL_URI,
    events: {
		maxExpandCount: 10,
    }
};
```

> This situation does not exist. When we query CalDav with an end date specified, the response we get only includes recurring events that start (or have an occurrence) between the startRange and endRange.

Ok, I didn't know that. Thanks for clarifying. You could add a comment in the code that explains this.

About the logic behind `maxExpandCount`:

I think `maxExpandCount` should be a finite number by default and should be applied to all situations, even when there's a start and an end date. Specifying `maxExpandCount` gives you the guarantee that scrapegoat won't expand events above a given limit, even when the end date is far in the future. Personally I think using `Infinity` for "Don't apply any limit" is ok, but I'm also ok with using `undefined`. In both cases, it needs to be explained in the README. I would use a finite number by default because often developers don't think about these edge cases. But I think the number can be really high, like 5.000 (this would include a reoccurring event every day for 10 years).

Regarding your concern when an `endDate` is missing: I think it's ok to have the possibility for an infinite loop. Since the default for `maxExpandCount` is a finite number, you need disable this safety net explicitly. If you do that, you're on your own. Again, I think there should be a warning in the README.

Reply:
Cool, thanks for the clarification. This is now implemented and ready for review.